misc/metrics/src/identify.rs

```diff
@@ -213,55 +175,104 @@ struct AddressLabels {
-    fn encode(&self, mut encoder: MetricEncoder) -> Result<(), std::fmt::Error> {
-        let count_by_protocol = self
-            .peers
-            .lock()
-            .expect("Lock not to be poisoned")
-            .iter()
-            .fold(
-                HashMap::<String, i64>::default(),
-                |mut acc, (_, protocols)| {
-                    for protocol in protocols {
-                        let count = acc.entry(protocol.to_string()).or_default();
-                        *count += 1;
-                    }
-                    acc
-                },
-            );
+impl Collector for Peers {
+    fn collect<'a>(
+        &'a self,
+    ) -> Box<dyn Iterator<Item = (Cow<'a, Descriptor>, MaybeOwned<'a, Box<dyn LocalMetric>>)> + 'a>
+    {
+        let mut count_by_protocols: HashMap<String, i64> = Default::default();
+        let mut count_by_listen_addresses: HashMap<String, i64> = Default::default();
+        let mut count_by_observed_addresses: HashMap<String, i64> = Default::default();
+
+        for (_, peer_info) in self.0.lock().unwrap().iter() {
+            {
+                let mut protocols: Vec<_> = peer_info
+                    .protocols
+                    .iter()
+                    .map(|p| {
+                        if ALLOWED_PROTOCOLS.contains(&p.as_bytes()) {
+                            p.to_string()
+                        } else {
+                            "unrecognized".to_string()
+                        }
+                    })
+                    .collect();
+                protocols.sort();
+                protocols.dedup();
+
+                for protocol in protocols.into_iter() {
+                    let count = count_by_protocols.entry(protocol).or_default();
+                    *count += 1;
+                }
+            }
+
+            {
+                let mut addrs: Vec<_> = peer_info
+                    .listen_addrs
+                    .iter()
+                    .map(protocol_stack::as_string)
+                    .collect();
+                addrs.sort();
+                addrs.dedup();
+
+                for addr in addrs {
+                    let count = count_by_listen_addresses.entry(addr).or_default();
+                    *count += 1;
+                }
+            }
 
-        for (protocol, count) in count_by_protocol {
-            encoder
-                .encode_family(&[("protocol", protocol)])?
-                .encode_gauge(&count)?;
+            {
+                let count = count_by_observed_addresses
+                    .entry(protocol_stack::as_string(&peer_info.observed_addr))
+                    .or_default();
+                *count += 1;
+            }
         }
 
-        Ok(())
-    }
+        let count_by_protocols: Box<dyn LocalMetric> =
+            Box::new(RefCell::new(count_by_protocols.into_iter().map(
+                |(protocol, count)| ([("protocol", protocol)], ConstGauge::new(count)),
+            )));
```

Comment:
See https://github.com/prometheus/client_rust/pull/82/files#r1158205762. I've found it very hard to discover what happens here.

Reply:
This pull request is now updated to the proposal in https://github.com/prometheus/client_rust/pull/138.